File: Data-structure/trie.py
class Node:
    def __init__(self, letter):
        self.letter = letter
        self.childs = list()
        self.terminal = 0
        self.count = 0


class Trie:
    def __init__(self):
        self.root = Node('')
        self.curr = self.root

    def __insert_letter(self, head, letter):
        for child in head.childs:
            if child.letter == letter:
                return child
        else:
            n = Node(letter)
            head.childs.append(n)
            return n

    def insert(self, word):
        self.curr = self.root
        for letter in word:
            self.curr = self.__insert_letter(self.curr,letter)
        self.curr.terminal = True
        self.curr.count += 1


    def has_word(self, word):
        self.curr = self.root
        for letter in word:
            for child in self.curr.childs:
                if child.letter == letter:
                    self.curr = child
                    break
            else:
                return False
        if self.curr.terminal:
            return True
        else:
            return False

    def count(self, word):
        self.curr = self.root
        for letter in word:
            for child in self.curr.childs:
                if child.letter == letter:
                    self.curr = child
                    break
            else:
                return 0
        if self.curr.terminal:
            return self.curr.count
        else:
            return 0
            
                

t = Trie()
t.insert("shohag")   
t.insert("shohag")
t.insert("shohag")
print(t.has_word("shohag"))
x = t.count('shohag')
print(x)

File: Data-structure/trie.cpp
#include <bits/stdc++.h>
using namespace std;

class TrieNode {
  private:
    bool endMark;
    TrieNode* next[29];
    
  public:
    bool isLetterExist(char letter) {
        int id = letter - 'a';
        return next[id] != NULL;
    }
    
    void insert(char letter, TrieNode* node) {
        int id = letter - 'a';
        if (next[id] == NULL) {
            next[id] = node;
        }
    }
    
    TrieNode* getNode(char letter) {
        int id = letter - 'a';
        return next[id];
    }
    
    void setEndMark(bool flag) {
        endMark = flag;
    }
    
    bool getEndMark() {
        return endMark;
    }
};

class Trie {
private:
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* curr = root;
        for (int i = 0; i < word.size(); i++) {
            if (!curr->isLetterExist(word[i])) {
                curr->insert(word[i], new TrieNode());
            }
            curr = curr->getNode(word[i]);
        }
        curr->setEndMark(true);
    }
    
    bool search(string word) {
        TrieNode* curr = root;
        for (int i = 0; i < word.size(); i++) {
            if (!curr->isLetterExist(word[i])) {
                return false;
            }
            curr = curr->getNode(word[i]);
        }
        return curr->getEndMark();
    }
    
    bool startsWith(string prefix) {
        TrieNode* curr = root;
        for (int i = 0; i < prefix.size(); i++) {
            if (!curr->isLetterExist(prefix[i])) {
                return false;
            }
            curr = curr->getNode(prefix[i]);
        }
        return true;
    }
};

Trie* obj = new Trie();
string word = "love";
string prefix = "lov";
obj->insert(word);
bool param_2 = obj->search(word);
bool param_3 = obj->startsWith(prefix);